apps/remote_control/lib/lexical/remote_control/code_intelligence/rename.ex

```diff
@@ -0,0 +1,178 @@
+      {:ok, {module_or_struct, module}, range} when module_or_struct in [:struct, :module] ->
+        {:ok, module, range}
+
+      {:ok, other, _} ->
+        {:error, {:unsupported_entity, other}}
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  defp search_related_candidates(document, position, entity, range) do
+    cursor_entity_string = cursor_entity_string(range)
+
+    entities =
+      exacts(entity, cursor_entity_string)
+
+    # Users won't always want to rename descendants of a module.
+    # For instance, when there are no more submodules after the cursor.
+    # like: `defmodule TopLevel.Mo|dule do`
+    # in most cases, users only want to rename the module itself.
+    #
+    # However, there's an exception when the cursor is in the middle,
+    # such as `Top.Mo|dule.ChildModule`. If we rename it to `Top.Renamed.Child`,
+    # it would be natural to also rename `Module.ChildModule` to `Renamed.Child`.
+    if at_the_middle_of_module?(document, position, range) do
+      entities ++ descendants(entity, cursor_entity_string)
+    else
+      entities
+    end
+  end
+
+  defp at_the_middle_of_module?(document, position, range) do
+    range_text = range_text(range)
+
+    case Ast.surround_context(document, position) do
+      {:ok, %{context: {:alias, alias}}} ->
+        String.length(range_text) < length(alias)
+
+      _ ->
+        false
+    end
+  end
+
+  defp cursor_entity_string(range) do
+    # Parent.|Module -> Module
+    range
+    |> range_text()
+    |> String.split(".")
+    |> List.last()
+  end
+
+  defp descendants(entity, cursor_entity_string) do
+    entity_string = inspect(entity)
+    prefix = "#{entity_string}."
+
+    case Store.prefix(prefix, subject: [:definition, :reference]) do
+      {:ok, results} ->
+        filtered =
+          results
+          |> Enum.filter(fn result ->
+            range_text = range_text(result.range)
+            String.contains?(range_text, cursor_entity_string)
+          end)
+          |> adjust_range(entity)
+
+        filtered
+
+      _ ->
+        []
+    end
+  end
+
+  defp exacts(entity, cursor_entity_string) do
+    entity_string = inspect(entity)
+
+    case Store.exact(entity_string, subject: [:definition, :reference]) do
+      {:ok, results} ->
+        filtered =
+          Enum.filter(results, fn result ->
+            range_text = range_text(result.range)
+            String.contains?(range_text, cursor_entity_string)
+          end)
+
+        filtered
+
+      _ ->
+        []
+    end
+  end
+
+  defp adjust_range(entries, entity) do
+    for entry <- entries do
+      location = {entry.range.start.line, entry.range.start.character}
+      uri = Document.Path.ensure_uri(entry.path)
+
+      case resolve_entity_range(uri, location, entity) do
+        {:ok, range} ->
+          %{entry | range: range}
+
+        :error ->
+          :error
+      end
+    end
+    |> Enum.reject(&(&1 == :error))
+  end
```

Comment:
Two suggestions here:

1. I don't think `location` serves any purpose -- `resolve_entity_range` can just take `entry.range.start`.
2. This can be cleaned up nicely by using the filtering feature of `for`.

```suggestion
  defp adjust_range(entries, entity) do
    for entry <- entries,
        uri = Document.Path.ensure_uri(entry.path),
        {:ok, range} = resolve_entity_range(uri, entry.range.start, entity) do
      %{entry | range: range}
    end
  end
```

Reply:
good catch, thanks.